pkg/descheduler/strategies/node_affinity.go

```diff
@@ -21,16 +21,70 @@ import (
 	"fmt"
 
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog/v2"
-
 	"sigs.k8s.io/descheduler/pkg/api"
 	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
 	nodeutil "sigs.k8s.io/descheduler/pkg/descheduler/node"
 	podutil "sigs.k8s.io/descheduler/pkg/descheduler/pod"
 	"sigs.k8s.io/descheduler/pkg/utils"
 )
 
+const NodeAffinityName = "RemovePodsViolatingNodeAffinity"
+
+// NewRemovePodsViolatingNodeAffinity returns a StrategyController to run Node affinity with informers
+func NewRemovePodsViolatingNodeAffinity(
+	ctx context.Context,
+	client clientset.Interface,
+	sharedInformerFactory informers.SharedInformerFactory,
+	strategy api.DeschedulerStrategy,
+	podEvictor *evictions.PodEvictor,
+	nodeSelector string,
+	stopChannel chan struct{},
+) *StrategyController {
+	c := &StrategyController{
+		ctx:                   ctx,
+		client:                client,
+		queue:                 workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		sharedInformerFactory: sharedInformerFactory,
+		strategy:              strategy,
+		podEvictor:            podEvictor,
+		nodeSelector:          nodeSelector,
+		stopChannel:           stopChannel,
+		name:                  NodeAffinityName,
+	}
+	sharedInformerFactory.Core().V1().Nodes().Informer().AddEventHandler(nodeEventHandler(c))
+	return c
+}
+
+func nodeEventHandler(c *StrategyController) cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			nodes, err := nodeutil.ReadyNodes(c.ctx, c.client, c.sharedInformerFactory.Core().V1().Nodes(), c.nodeSelector)
```

Comment:
If strategies run serially, the time between listing nodes and running strategy may be long, if another event comes during this period, we'll list nodes again, acturally we only need to do this once. So like what was disscussed in #469 , listing and filtering nodes could be done in each strategies, we can also custom `nodeSelector` for each strategies by doing so.

Reply:
> if another event comes during this period, we'll list nodes again, acturally we only need to do this once

I'm not sure, because the point of reacting to Node events as they happen is to get a real-time updated view of the cluster to operate on. So each event should trigger a re-list so the strategy doesn't have old data.

I do agree though with refactoring this a bit like was previously mentioned.

Reply:
All the "reactive" strategies listening to node informers need to take into account strategies like LowNodeUtilization. `LowNodeUtilization` needs to take into account entire cluster (or its reasonable subset). So if the strategies are ran in incorrect order, e.g. `LowNodeUtilization` followed by running `PodLifeTime` removing many old pods, running `LowNodeUtilization` before might be actually nullified and make the overall resource consumption worst. So far, the order of strategies was [kinda hardcoded](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/descheduler.go#L72-L81) (depending on the map iterator). We might compute some static impact/score of each strategy to overall resource utilization and run the strategies on order starting with strategies changing utilization the most to strategies changing it the least (just a though, it might not be possible).

Reply:
I doubt that `LowNodeUtilization` will be able to easily convert to be reactive. That is the point of the design I proposed here, that we can begin to convert some of the strategies while keeping the hardcoded order of others.

I think @lixiang233's idea of a single strategy controller with a master registry of event handlers would solve any ordering issues. That with a mutex ensures we aren't running 2 strategies at the same time.

For example, with an interval of 60 minutes we could have:
```
0min - interval run of strategies
12 min - strategycontroller runs nodeaffinity
60 min - interval run
119 min - strategycontroller run (mutex lock)
121 min - interval (blocked waiting for mutex) runs
181 min - interval run
```

And since each periodic interval does a node re-list, those are working with an up-to-date list each time. Any results from NodeAffinity/NodeTaints won't interfere. So really, we are just triggering a descheduling run at dynamic intervals along with the hardcoded/cron intervals